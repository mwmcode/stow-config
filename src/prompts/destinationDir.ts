import { Input } from '@cliffy/prompt';
import { transformInput } from './helpers/transformInput.ts';
import { confirmCreate } from './confirmCreate.ts';

export async function promptDestinationDir() {
  return await Input.prompt({
    files: true,
    message: 'Enter path to `destination` directory',
    hint: 'Location to move congis to?',
    transform: transformInput,
    validate: async value => {
      const destDir = transformInput(value);
      try {
        return Deno.lstatSync(destDir).isDirectory;
      } catch (error) {
        if (error instanceof Deno.errors.NotFound) {
          return await confirmCreate(destDir);
        } else {
          console.error(`\n${error}`);
          Deno.exit(1);
        }
      }
    },
  });
}
