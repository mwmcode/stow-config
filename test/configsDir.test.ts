import { snapshotTest } from '@cliffy/testing';
import { ansi } from '@cliffy/ansi';
import { promptConfigDir } from '../src/prompts/index.ts';
import { join } from '@std/path';
import {
  TESTSC_HOME_ALTERNATIVE,
  TEST_DENO_ARGS,
  TestDirectories,
} from '@/consts.ts';
import { promptDestinationDir } from '@/src/prompts/destinationDir.ts';

await snapshotTest({
  name: 'Prompt user to use default `$HOME/.config`',
  meta: import.meta,
  stdin: ansi.cursorForward.text('\n').toArray(),
  denoArgs: TEST_DENO_ARGS,
  async fn() {
    Deno.env.set(TESTSC_HOME_ALTERNATIVE, join(Deno.cwd(), 'test/data'));
    const answer = await promptConfigDir();
    console.log({ answer });
    Deno.env.delete(TESTSC_HOME_ALTERNATIVE);
  },
});

await snapshotTest({
  name: 'Ask user to enter path to configs dir',
  meta: import.meta,
  stdin: ansi
    .text(join(Deno.cwd(), TestDirectories.configSrcDir))
    .text('\n')
    .toArray(),
  denoArgs: TEST_DENO_ARGS,
  async fn() {
    Deno.env.set(TESTSC_HOME_ALTERNATIVE, 'WRONG_DIR');
    const answer = await promptConfigDir();
    console.log({ answer });
  },
});

await snapshotTest({
  name: 'Ask user to enter path to target dir',
  meta: import.meta,
  stdin: ansi
    .text(join(Deno.cwd(), TestDirectories.configDestDir))
    .text('\n')
    .toArray(),
  denoArgs: ['-R', '--allow-env=SNAPSHOT_TEST_NAME'],
  async fn() {
    Deno.env.set(TESTSC_HOME_ALTERNATIVE, TESTSC_HOME_ALTERNATIVE);
    const answer = await promptDestinationDir();
    console.log({ answer });
  },
});
