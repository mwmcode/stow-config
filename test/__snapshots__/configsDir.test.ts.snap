export const snapshot = {};

snapshot[`Prompt user to use default \`\$HOME/.config\` 1`] = `
stdout:
'{ testHomeDir: "/Users/mch/Developer/deno/stow-config/test/data" }
? Use ðŸ“‚ /Users/mch/Developer/deno/stow-config/test/data to search for configs? â€º No / Yes\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Use ðŸ“‚ /Users/mch/Developer/deno/stow-config/test/data to search for configs? â€º No / Yes\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Use ðŸ“‚ /Users/mch/Developer/deno/stow-config/test/data to search for configs? â€º Yes
\\x1b[?25h\\x1b[?25h{ answer: "/Users/mch/Developer/deno/stow-config/test/data" }
'
stderr:
""
`;

snapshot[`Ask user to enter path to configs dir 1`] = `
stdout:
'{ testHomeDir: "WRONG_DIR" }
? Path path to \`configs\` directory â€º \\x1b[38G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch\\x1b[46G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer\\x1b[54G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno\\x1b[62G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno/stow-config\\x1b[70G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno/stow-config/test\\x1b[78G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno/stow-config/test/data/config-src\\x1b[86G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno/stow-config/test/data/config-src\\x1b[94G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno/stow-config/test/data/config-src\\x1b[96G\\x1b[G\\x1b[0J
? Path path to \`configs\` directory â€º /Users/mch/Developer/deno/stow-config/test/data/config-src
\\x1b[?25h\\x1b[?25h{
  answer: "/Users/mch/Developer/deno/stow-config/test/data/config-src"
}
'
stderr:
""
`;
